File: 300/337 - House Robber III.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int rob(TreeNode root) {
        int[] ret = dfs(root);
        
        return Math.max(ret[0], ret[1]);
    }
    
    private int[] dfs(TreeNode root) {
        int[] ret = {0, 0}, l, r;
        
        if (root == null)
            return ret;
        
        l = dfs(root.left);
        r = dfs(root.right);
        ret[0] = Math.max(l[0], l[1]) + Math.max(r[0], r[1]);
        ret[1] = l[0] + r[0] + root.val;
        
        return ret;
    }
}

File: csharp/337.house-robber-iii.cs
/*
 * @lc app=leetcode id=337 lang=csharp
 *
 * [337] House Robber III
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {

    private (int, int) RobRecurse(TreeNode root) {
        if (root == null) {
            return (0, 0);
        }
        var left = RobRecurse(root.left);
        var right = RobRecurse(root.right);
        int noRob = Math.Max(left.Item1, left.Item2) + Math.Max(right.Item1, right.Item2), rob = root.val + left.Item1 + right.Item1;
        return (noRob, rob);
    }

    public int Rob(TreeNode root) {
        var ret = RobRecurse(root);
        return Math.Max(ret.Item1, ret.Item2);
    }
}

